<a href="https://colab.research.google.com/github/ev-greshkina/DataAnalyst-educational-course/blob/main/%D0%98%D1%82%D0%BE%D0%B3%D0%BE%D0%B2%D0%B0%D1%8F_%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D0%B0_%D0%90%D0%B2%D0%B8%D0%B0%D1%81%D0%BE%D0%BE%D0%B1%D1%89%D0%B5%D0%BD%D0%B8%D0%B5_%D0%A1%D0%A8%D0%90.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Шаг 1: Загрузка и предобработка данных

In [1]:
import pandas as pd

# Загрузка данных
flights = pd.read_csv('flights.csv')
airports = pd.read_csv('airports.csv')
airlines = pd.read_csv('airlines.csv')


<ipython-input-1-ba7857dc2c86>:4: DtypeWarning: Columns (7,8) have mixed types. Specify dtype option on import or set low_memory=False.
  flights = pd.read_csv('flights.csv')


**Задача 1: Найти аэропорт с минимальной задержкой вылета**

In [ ]:
# Создать копию DataFrame 'flights' для выполнения преобразований и вычислений с копией
# а не с оригиналом
flights_task_1 = flights

In [ ]:
# Преобразование задержки в числовой формат
flights_task_1['DEPARTURE_DELAY'] = pd.to_numeric(flights_task_1['DEPARTURE_DELAY'], errors='coerce')

# Удаляем строки с пустыми значениями задержек
flights_task_1 = flights_task_1.dropna(subset=['DEPARTURE_DELAY', 'ORIGIN_AIRPORT'])

flights_task_1

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AA           2336      N3KUAA   
2        2015      1    1            4      US            840      N171US   
3        2015      1    1            4      AA            258      N3HYAA   
4        2015      1    1            4      AS            135      N527AS   
...       ...    ...  ...          ...     ...            ...         ...   
5819074  2015     12   31            4      B6            688      N657JB   
5819075  2015     12   31            4      B6            745      N828JB   
5819076  2015     12   31            4      B6           1503      N913JB   
5819077  2015     12   31            4      B6            333      N527JB   
5819078  2015     12   31            4      B6            839      N534JB   

        ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  \
0                  ANC                 SEA                    5  ...   
1                  LAX                 PBI                   10  ...   
2                  SFO                 CLT                   20  ...   
3                  LAX                 MIA                   20  ...   
4                  SEA                 ANC                   25  ...   
...                ...                 ...                  ...  ...   
5819074            LAX                 BOS                 2359  ...   
5819075            JFK                 PSE                 2359  ...   
5819076            JFK                 SJU                 2359  ...   
5819077            MCO                 SJU                 2359  ...   
5819078            JFK                 BQN                 2359  ...   

         ARRIVAL_TIME  ARRIVAL_DELAY  DIVERTED  CANCELLED  \
0               408.0          -22.0         0          0   
1               741.0           -9.0         0          0   
2               811.0            5.0         0          0   
3               756.0           -9.0         0          0   
4               259.0          -21.0         0          0   
...               ...            ...       ...        ...   
5819074         753.0          -26.0         0          0   
5819075         430.0          -16.0         0          0   
5819076         432.0           -8.0         0          0   
5819077         330.0          -10.0         0          0   
5819078         442.0            2.0         0          0   

         CANCELLATION_REASON  AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  \
0                        NaN               NaN             NaN            NaN   
1                        NaN               NaN             NaN            NaN   
2                        NaN               NaN             NaN            NaN   
3                        NaN               NaN             NaN            NaN   
4                        NaN               NaN             NaN            NaN   
...                      ...               ...             ...            ...   
5819074                  NaN               NaN             NaN            NaN   
5819075                  NaN               NaN             NaN            NaN   
5819076                  NaN               NaN             NaN            NaN   
5819077                  NaN               NaN             NaN            NaN   
5819078                  NaN               NaN             NaN            NaN   

         LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0                        NaN            NaN  
1                        NaN            NaN  
2                        NaN            NaN  
3                        NaN            NaN  
4                        NaN            NaN  
...                      ...            ...  
5819074                  NaN            NaN  
5819075                  NaN            NaN  
5819076                  NaN            NaN  
5819077                  NaN            NaN  
5819078

In [ ]:
# В столбце ORIGIN_AIRPORT имеются цифровые ITA-CODE. Цифровые обозначения
# вероятно являются ошибочно внесенными данныи, поэтому строки содержащие цифры в
# столбцах ORIGIN_AIRPORT и DESTINATION_AIRPORT подлежат удалению. Идентифицировать аэропрт
# по этим данным не возможно
# Удаляем строки, содержащие цифры

# Преобразовать данные в столбце ORIGIN_AIRPORT в строковые значения
flights_task_1['ORIGIN_AIRPORT'] = flights_task_1['ORIGIN_AIRPORT'].astype(str)

# Фильтрация строк, содержащих цифры
flights_task_1 = flights_task_1[~flights_task_1['ORIGIN_AIRPORT'].str.contains('[0-9]')]
flights_task_1

<ipython-input-4-cab9bdc3b79b>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  flights_task_1['ORIGIN_AIRPORT'] = flights_task_1['ORIGIN_AIRPORT'].astype(str)


YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AA           2336      N3KUAA   
2        2015      1    1            4      US            840      N171US   
3        2015      1    1            4      AA            258      N3HYAA   
4        2015      1    1            4      AS            135      N527AS   
...       ...    ...  ...          ...     ...            ...         ...   
5819074  2015     12   31            4      B6            688      N657JB   
5819075  2015     12   31            4      B6            745      N828JB   
5819076  2015     12   31            4      B6           1503      N913JB   
5819077  2015     12   31            4      B6            333      N527JB   
5819078  2015     12   31            4      B6            839      N534JB   

        ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  \
0                  ANC                 SEA                    5  ...   
1                  LAX                 PBI                   10  ...   
2                  SFO                 CLT                   20  ...   
3                  LAX                 MIA                   20  ...   
4                  SEA                 ANC                   25  ...   
...                ...                 ...                  ...  ...   
5819074            LAX                 BOS                 2359  ...   
5819075            JFK                 PSE                 2359  ...   
5819076            JFK                 SJU                 2359  ...   
5819077            MCO                 SJU                 2359  ...   
5819078            JFK                 BQN                 2359  ...   

         ARRIVAL_TIME  ARRIVAL_DELAY  DIVERTED  CANCELLED  \
0               408.0          -22.0         0          0   
1               741.0           -9.0         0          0   
2               811.0            5.0         0          0   
3               756.0           -9.0         0          0   
4               259.0          -21.0         0          0   
...               ...            ...       ...        ...   
5819074         753.0          -26.0         0          0   
5819075         430.0          -16.0         0          0   
5819076         432.0           -8.0         0          0   
5819077         330.0          -10.0         0          0   
5819078         442.0            2.0         0          0   

         CANCELLATION_REASON  AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  \
0                        NaN               NaN             NaN            NaN   
1                        NaN               NaN             NaN            NaN   
2                        NaN               NaN             NaN            NaN   
3                        NaN               NaN             NaN            NaN   
4                        NaN               NaN             NaN            NaN   
...                      ...               ...             ...            ...   
5819074                  NaN               NaN             NaN            NaN   
5819075                  NaN               NaN             NaN            NaN   
5819076                  NaN               NaN             NaN            NaN   
5819077                  NaN               NaN             NaN            NaN   
5819078                  NaN               NaN             NaN            NaN   

         LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0                        NaN            NaN  
1                        NaN            NaN  
2                        NaN            NaN  
3                        NaN            NaN  
4                        NaN            NaN  
...                      ...            ...  
5819074                  NaN            NaN  
5819075                  NaN            NaN  
5819076                  NaN            NaN  
5819077                  NaN            NaN  
5819078

In [ ]:
# Группировка по аэропорту и расчет средней задержки вылета
airport_delay = flights_task_1.groupby('ORIGIN_AIRPORT')['DEPARTURE_DELAY'].mean().reset_index()
airport_delay

ORIGIN_AIRPORT  DEPARTURE_DELAY
0              ABE         8.844812
1              ABI         5.783989
2              ABQ         8.877015
3              ABR        10.101056
4              ABY         8.880046
..             ...              ...
317            WRG         3.091884
318            WYS         1.375000
319            XNA        11.886184
320            YAK        -6.292237
321            YUM         1.232470

[322 rows x 2 columns]

In [ ]:
# Таблица полученных значений содержит отрицательные значения, которые вероятно
# означают, что самолет вылетел с опережением графика. Для получения данных о наименьшей
# задержке вылета требуется удалить все строки с отрицательным значением DEPARTURE DELAY
airport_delay = airport_delay.drop(airport_delay.index[airport_delay['DEPARTURE_DELAY'] < 0])

# Отортировать и вывести полученные значения
sorted = airport_delay.sort_values(by='DEPARTURE_DELAY')
sorted

ORIGIN_AIRPORT  DEPARTURE_DELAY
274            SCC         0.154122
10             ADQ         0.405034
184            LIH         0.556254
89             DLG         0.558442
46             BTM         0.674383
..             ...              ...
231            OTH        17.777358
293            STC        18.692308
147            HYA        23.182927
219            MVY        25.907317
154            ILG        29.391753

[311 rows x 2 columns]

In [ ]:
# Поиск аэропорта с минимальной задержкой
# Если значение задержки вылета отрицательное, то это означает, что вылет выполнен с опережением графика
min_delay_airport = airport_delay.loc[airport_delay['DEPARTURE_DELAY'].idxmin()]
min_delay_airport

ORIGIN_AIRPORT          SCC
DEPARTURE_DELAY    0.154122
Name: 274, dtype: object

In [ ]:
# Получение информации о названии аэропорта
airport_info = airports[airports['IATA_CODE'] == min_delay_airport['ORIGIN_AIRPORT']]
airport_info

IATA_CODE                                  AIRPORT       CITY STATE  \
274       SCC  Deadhorse Airport (Prudhoe Bay Airport)  Deadhorse    AK   

    COUNTRY  LATITUDE  LONGITUDE  
274     USA  70.19476 -148.46516

In [ ]:
# Вывод результата
print(f"Минимальная средняя задержка вылета: {airport_info['AIRPORT'].values[0]} ({min_delay_airport['ORIGIN_AIRPORT']}) с задержкой {min_delay_airport['DEPARTURE_DELAY']:.2f} минут")

Минимальная средняя задержка вылета: Deadhorse Airport (Prudhoe Bay Airport) (SCC) с задержкой 0.15 минут


**Задача 2: Самая пунктуальная авиакомпания на прилет в Los Angeles International Airport**

In [ ]:
# Создать копию DataFrame 'flights' для выполнения преобразований и вычислений с копией
# а не с оригиналом
flights_task_2 = flights

In [ ]:
# Преобразование задержки в числовой формат
flights_task_2['ARRIVAL_DELAY'] = pd.to_numeric(flights_task_2['ARRIVAL_DELAY'], errors='coerce')


In [ ]:
# Удаление строк с пустыми значениями в столбцах 'ARRIVAL_DELAY' и 'DESTINATION_AIRPORT'
flights_task_2 = flights_task_2.dropna(subset=['ARRIVAL_DELAY', 'DESTINATION_AIRPORT'])

In [ ]:
# Группировка по авиакомпаниям и расчет средней задержки для рейсов в Лос-Анджелес
la_flights = flights_task_2[flights_task_2['DESTINATION_AIRPORT'] == 'LAX']

In [ ]:
# Проверка, есть ли данные для анализа
if not la_flights.empty:
    airline_punctuality = la_flights.groupby('AIRLINE')['ARRIVAL_DELAY'].mean().reset_index()

    # Поиск самой пунктуальной авиакомпании
    if not airline_punctuality.empty:
        most_punctual_airline = airline_punctuality.loc[airline_punctuality['ARRIVAL_DELAY'].idxmin()]

        # Получение информации о названии авиакомпании
        airline_info = airlines[airlines['IATA_CODE'] == most_punctual_airline['AIRLINE']]

        if not airline_info.empty:
            print(f"Самая пунктуальная авиакомпания на прилет в LAX: {airline_info['AIRLINE'].values[0]} с задержкой {most_punctual_airline['ARRIVAL_DELAY']:.2f} минут")
        else:
            print("Информация об авиакомпании не найдена.")
    else:
        print("Нет данных для анализа пунктуальности авиакомпаний.")
else:
    print("Нет рейсов, прилетающих в LAX для анализа.")

Самая пунктуальная авиакомпания на прилет в LAX: Hawaiian Airlines Inc. с задержкой -4.67 минут


**Задача 3: Найти аэропорт, где самолёты проводят больше всего времени на рулении (среднее значение)**

In [ ]:
# Создать копию DataFrame 'flights' для выполнения преобразований и вычислений с копией
# а не с оригиналом
flights_task_3 = flights

In [ ]:
# Преобразование времени руления в числовой формат
flights_task_3['TAXI_OUT'] = pd.to_numeric(flights_task_3['TAXI_OUT'], errors='coerce')


In [ ]:
# Удаляем строки с пустыми значениями столбца ORIGIN_AIRPORT
flights_task_3 = flights_task_3.dropna(subset=['ORIGIN_AIRPORT'])

flights_task_3

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AA           2336      N3KUAA   
2        2015      1    1            4      US            840      N171US   
3        2015      1    1            4      AA            258      N3HYAA   
4        2015      1    1            4      AS            135      N527AS   
...       ...    ...  ...          ...     ...            ...         ...   
5819074  2015     12   31            4      B6            688      N657JB   
5819075  2015     12   31            4      B6            745      N828JB   
5819076  2015     12   31            4      B6           1503      N913JB   
5819077  2015     12   31            4      B6            333      N527JB   
5819078  2015     12   31            4      B6            839      N534JB   

        ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  \
0                  ANC                 SEA                    5  ...   
1                  LAX                 PBI                   10  ...   
2                  SFO                 CLT                   20  ...   
3                  LAX                 MIA                   20  ...   
4                  SEA                 ANC                   25  ...   
...                ...                 ...                  ...  ...   
5819074            LAX                 BOS                 2359  ...   
5819075            JFK                 PSE                 2359  ...   
5819076            JFK                 SJU                 2359  ...   
5819077            MCO                 SJU                 2359  ...   
5819078            JFK                 BQN                 2359  ...   

         ARRIVAL_TIME  ARRIVAL_DELAY  DIVERTED  CANCELLED  \
0               408.0          -22.0         0          0   
1               741.0           -9.0         0          0   
2               811.0            5.0         0          0   
3               756.0           -9.0         0          0   
4               259.0          -21.0         0          0   
...               ...            ...       ...        ...   
5819074         753.0          -26.0         0          0   
5819075         430.0          -16.0         0          0   
5819076         432.0           -8.0         0          0   
5819077         330.0          -10.0         0          0   
5819078         442.0            2.0         0          0   

         CANCELLATION_REASON  AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  \
0                        NaN               NaN             NaN            NaN   
1                        NaN               NaN             NaN            NaN   
2                        NaN               NaN             NaN            NaN   
3                        NaN               NaN             NaN            NaN   
4                        NaN               NaN             NaN            NaN   
...                      ...               ...             ...            ...   
5819074                  NaN               NaN             NaN            NaN   
5819075                  NaN               NaN             NaN            NaN   
5819076                  NaN               NaN             NaN            NaN   
5819077                  NaN               NaN             NaN            NaN   
5819078                  NaN               NaN             NaN            NaN   

         LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0                        NaN            NaN  
1                        NaN            NaN  
2                        NaN            NaN  
3                        NaN            NaN  
4                        NaN            NaN  
...                      ...            ...  
5819074                  NaN            NaN  
5819075                  NaN            NaN  
5819076                  NaN            NaN  
5819077                  NaN            NaN  
5819078

In [ ]:
# В столбце ORIGIN_AIRPORT имеются цифровые ITA-CODE. Цифровые обозначения
# вероятно являются ошибочно внесенными данныи, поэтому строки содержащие цифры в
# столбцах ORIGIN_AIRPORT и DESTINATION_AIRPORT подлежат удалению. Идентифицировать аэропрт
# по этим данным не возможно
# Удаляем строки, содержащие цифры

# Преобразовать данные в столбце ORIGIN_AIRPORT в строковые значения
flights_task_3['ORIGIN_AIRPORT'] = flights_task_3['ORIGIN_AIRPORT'].astype(str)

# Фильтрация строк, содержащих цифры
flights_task_3 = flights_task_3[~flights_task_3['ORIGIN_AIRPORT'].str.contains('[0-9]')]
flights_task_3

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AA           2336      N3KUAA   
2        2015      1    1            4      US            840      N171US   
3        2015      1    1            4      AA            258      N3HYAA   
4        2015      1    1            4      AS            135      N527AS   
...       ...    ...  ...          ...     ...            ...         ...   
5819074  2015     12   31            4      B6            688      N657JB   
5819075  2015     12   31            4      B6            745      N828JB   
5819076  2015     12   31            4      B6           1503      N913JB   
5819077  2015     12   31            4      B6            333      N527JB   
5819078  2015     12   31            4      B6            839      N534JB   

        ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  \
0                  ANC                 SEA                    5  ...   
1                  LAX                 PBI                   10  ...   
2                  SFO                 CLT                   20  ...   
3                  LAX                 MIA                   20  ...   
4                  SEA                 ANC                   25  ...   
...                ...                 ...                  ...  ...   
5819074            LAX                 BOS                 2359  ...   
5819075            JFK                 PSE                 2359  ...   
5819076            JFK                 SJU                 2359  ...   
5819077            MCO                 SJU                 2359  ...   
5819078            JFK                 BQN                 2359  ...   

         ARRIVAL_TIME  ARRIVAL_DELAY  DIVERTED  CANCELLED  \
0               408.0          -22.0         0          0   
1               741.0           -9.0         0          0   
2               811.0            5.0         0          0   
3               756.0           -9.0         0          0   
4               259.0          -21.0         0          0   
...               ...            ...       ...        ...   
5819074         753.0          -26.0         0          0   
5819075         430.0          -16.0         0          0   
5819076         432.0           -8.0         0          0   
5819077         330.0          -10.0         0          0   
5819078         442.0            2.0         0          0   

         CANCELLATION_REASON  AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  \
0                        NaN               NaN             NaN            NaN   
1                        NaN               NaN             NaN            NaN   
2                        NaN               NaN             NaN            NaN   
3                        NaN               NaN             NaN            NaN   
4                        NaN               NaN             NaN            NaN   
...                      ...               ...             ...            ...   
5819074                  NaN               NaN             NaN            NaN   
5819075                  NaN               NaN             NaN            NaN   
5819076                  NaN               NaN             NaN            NaN   
5819077                  NaN               NaN             NaN            NaN   
5819078                  NaN               NaN             NaN            NaN   

         LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0                        NaN            NaN  
1                        NaN            NaN  
2                        NaN            NaN  
3                        NaN            NaN  
4                        NaN            NaN  
...                      ...            ...  
5819074                  NaN            NaN  
5819075                  NaN            NaN  
5819076                  NaN            NaN  
5819077                  NaN            NaN  
5819078

In [ ]:
# Группировка по аэропорту и расчет среднего времени руления
taxi_time = flights_task_3.groupby('ORIGIN_AIRPORT')['TAXI_OUT'].mean().reset_index()

taxi_time

ORIGIN_AIRPORT   TAXI_OUT
0              ABE  13.408949
1              ABI   9.188004
2              ABQ  11.987672
3              ABR  16.600302
4              ABY  12.084198
..             ...        ...
317            WRG   7.820827
318            WYS  11.903846
319            XNA  14.681277
320            YAK   7.919330
321            YUM  14.348436

[322 rows x 2 columns]

In [ ]:
# Поиск аэропорта с максимальным временем руления
max_taxi_time_airport = taxi_time.loc[taxi_time['TAXI_OUT'].idxmax()]

max_taxi_time_airport

ORIGIN_AIRPORT          JFK
TAXI_OUT          27.009783
Name: 166, dtype: object

In [ ]:
# Получение информации о названии аэропорта
max_airport_info = airports[airports['IATA_CODE'] == max_taxi_time_airport['ORIGIN_AIRPORT']]

max_airport_info

IATA_CODE                                            AIRPORT      CITY  \
166       JFK  John F. Kennedy International Airport (New Yor...  New York   

    STATE COUNTRY  LATITUDE  LONGITUDE  
166    NY     USA  40.63975  -73.77893

In [ ]:
# Вывод результата
print(f"Аэропорт с максимальным временем руления: {max_airport_info['AIRPORT'].values[0]} ({max_taxi_time_airport['ORIGIN_AIRPORT']}) с временем {max_taxi_time_airport['TAXI_OUT']:.2f} минут")

Аэропорт с максимальным временем руления: John F. Kennedy International Airport (New York International Airport) (JFK) с временем 27.01 минут


**Задача 4: Построить модель для выбора топ-3 аэропортов прилета**

In [2]:
# Подключить необходимые библиотеки
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
import numpy as np

In [3]:
# Преобразование задержки в числовой формат с оптимизацией памяти
flights['DEPARTURE_DELAY'] = pd.to_numeric(flights['DEPARTURE_DELAY'], errors='coerce', downcast='float')
flights['ARRIVAL_DELAY'] = pd.to_numeric(flights['ARRIVAL_DELAY'], errors='coerce', downcast='float')

In [4]:
# Убираем строки с отсутствующими значениями
flights = flights.dropna(subset=['DEPARTURE_DELAY', 'ARRIVAL_DELAY', 'ORIGIN_AIRPORT', 'DESTINATION_AIRPORT'])

In [5]:
# В столбце ORIGIN_AIRPORT имеются цифровые ITA-CODE. Цифровые обозначения
# вероятно являются ошибочно внесенными данныи, поэтому строки содержащие цифры в
# столбцах ORIGIN_AIRPORT и DESTINATION_AIRPORT подлежат удалению. Идентифицировать аэропрт
# по этим данным не возможно
# Удаляем строки, содержащие цифры

# Преобразовать данные в столбце ORIGIN_AIRPORT в строковые значения
flights['ORIGIN_AIRPORT'] = flights['ORIGIN_AIRPORT'].astype(str)

# Фильтрация строк, содержащих цифры
flights = flights[~flights['ORIGIN_AIRPORT'].str.contains('[0-9]')]

In [6]:
# Выполнить замену категориальных данных в столбце ORIGIN_AIRPORT
# Извлеч столбец ORIGIN_AIRPORT в отдельный Series
airport = flights.ORIGIN_AIRPORT

# Удалить дубликаты названий аэропортов
airport = airport.drop_duplicates()

# Сбросить параметры индексирования методом reset_index
airport = airport.reset_index()

# Удалить стобец index для т.к. сохранены исходыне индексы и они не по порядку.
# Сохранить только столбец ORIGIN_AIRPORT
airport = airport.ORIGIN_AIRPORT

# Сбросить значение index
airport = airport.reset_index()
airport


index ORIGIN_AIRPORT
0        0            ANC
1        1            LAX
2        2            SFO
3        3            SEA
4        4            LAS
..     ...            ...
317    317            WYS
318    318            DLG
319    319            AKN
320    320            GST
321    321            HYA

[322 rows x 2 columns]

In [7]:
# Произвести замену значения в столбце ORIGIN_AIRPORT дата фрейма flights на соответствующее значение
# из созданной таблицы числовых замен значения аэропорта
# Объединить таблицы методом merge
flights = flights.merge(airport, how='left', on='ORIGIN_AIRPORT', suffixes=('', '1'))
flights


YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AA           2336      N3KUAA   
2        2015      1    1            4      US            840      N171US   
3        2015      1    1            4      AA            258      N3HYAA   
4        2015      1    1            4      AS            135      N527AS   
...       ...    ...  ...          ...     ...            ...         ...   
5231125  2015     12   31            4      B6            688      N657JB   
5231126  2015     12   31            4      B6            745      N828JB   
5231127  2015     12   31            4      B6           1503      N913JB   
5231128  2015     12   31            4      B6            333      N527JB   
5231129  2015     12   31            4      B6            839      N534JB   

        ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  \
0                  ANC                 SEA                    5  ...   
1                  LAX                 PBI                   10  ...   
2                  SFO                 CLT                   20  ...   
3                  LAX                 MIA                   20  ...   
4                  SEA                 ANC                   25  ...   
...                ...                 ...                  ...  ...   
5231125            LAX                 BOS                 2359  ...   
5231126            JFK                 PSE                 2359  ...   
5231127            JFK                 SJU                 2359  ...   
5231128            MCO                 SJU                 2359  ...   
5231129            JFK                 BQN                 2359  ...   

         ARRIVAL_DELAY  DIVERTED  CANCELLED  CANCELLATION_REASON  \
0                -22.0         0          0                  NaN   
1                 -9.0         0          0                  NaN   
2                  5.0         0          0                  NaN   
3                 -9.0         0          0                  NaN   
4                -21.0         0          0                  NaN   
...                ...       ...        ...                  ...   
5231125          -26.0         0          0                  NaN   
5231126          -16.0         0          0                  NaN   
5231127           -8.0         0          0                  NaN   
5231128          -10.0         0          0                  NaN   
5231129            2.0         0          0                  NaN   

         AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  \
0                     NaN             NaN            NaN                  NaN   
1                     NaN             NaN            NaN                  NaN   
2                     NaN             NaN            NaN                  NaN   
3                     NaN             NaN            NaN                  NaN   
4                     NaN             NaN            NaN                  NaN   
...                   ...             ...            ...                  ...   
5231125               NaN             NaN            NaN                  NaN   
5231126               NaN             NaN            NaN                  NaN   
5231127               NaN             NaN            NaN                  NaN   
5231128               NaN             NaN            NaN                  NaN   
5231129               NaN             NaN            NaN                  NaN   

         WEATHER_DELAY  index  
0                  NaN      0  
1                  NaN      1  
2                  NaN      2  
3                  NaN      1  
4                  NaN      3  
...                ...    ...  
5231125            NaN      1  
5231126            NaN     38  
5231127            NaN     38  
5231128            NaN     20  
5231129            NaN     38  

[5231130 rows x 32 columns]

In [8]:
# Удалить столбец ORIGIN_AIRPORT
flights.drop('ORIGIN_AIRPORT', axis=1, inplace=True)

# Переименовать столбец index в ORIGIN_AIRPORT
flights = flights.rename(columns={'index': 'ORIGIN_AIRPORT'})

flights

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AA           2336      N3KUAA   
2        2015      1    1            4      US            840      N171US   
3        2015      1    1            4      AA            258      N3HYAA   
4        2015      1    1            4      AS            135      N527AS   
...       ...    ...  ...          ...     ...            ...         ...   
5231125  2015     12   31            4      B6            688      N657JB   
5231126  2015     12   31            4      B6            745      N828JB   
5231127  2015     12   31            4      B6           1503      N913JB   
5231128  2015     12   31            4      B6            333      N527JB   
5231129  2015     12   31            4      B6            839      N534JB   

        DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  DEPARTURE_TIME  ...  \
0                       SEA                    5          2354.0  ...   
1                       PBI                   10             2.0  ...   
2                       CLT                   20            18.0  ...   
3                       MIA                   20            15.0  ...   
4                       ANC                   25            24.0  ...   
...                     ...                  ...             ...  ...   
5231125                 BOS                 2359          2355.0  ...   
5231126                 PSE                 2359          2355.0  ...   
5231127                 SJU                 2359          2350.0  ...   
5231128                 SJU                 2359          2353.0  ...   
5231129                 BQN                 2359            14.0  ...   

         ARRIVAL_DELAY  DIVERTED  CANCELLED  CANCELLATION_REASON  \
0                -22.0         0          0                  NaN   
1                 -9.0         0          0                  NaN   
2                  5.0         0          0                  NaN   
3                 -9.0         0          0                  NaN   
4                -21.0         0          0                  NaN   
...                ...       ...        ...                  ...   
5231125          -26.0         0          0                  NaN   
5231126          -16.0         0          0                  NaN   
5231127           -8.0         0          0                  NaN   
5231128          -10.0         0          0                  NaN   
5231129            2.0         0          0                  NaN   

         AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  \
0                     NaN             NaN            NaN                  NaN   
1                     NaN             NaN            NaN                  NaN   
2                     NaN             NaN            NaN                  NaN   
3                     NaN             NaN            NaN                  NaN   
4                     NaN             NaN            NaN                  NaN   
...                   ...             ...            ...                  ...   
5231125               NaN             NaN            NaN                  NaN   
5231126               NaN             NaN            NaN                  NaN   
5231127               NaN             NaN            NaN                  NaN   
5231128               NaN             NaN            NaN                  NaN   
5231129               NaN             NaN            NaN                  NaN   

         WEATHER_DELAY  ORIGIN_AIRPORT  
0                  NaN               0  
1                  NaN               1  
2                  NaN               2  
3                  NaN               1  
4                  NaN               3  
...                ...             ...  
5231125            NaN               1  
5231126            NaN              38  
5231127            NaN              38  
5231128

In [9]:
# Создание исходных данных для построения модели
X = flights[['DEPARTURE_DELAY', 'ORIGIN_AIRPORT']]
y = flights['ARRIVAL_DELAY']

In [10]:
%%time
# Разделение на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

CPU times: user 1.04 s, sys: 4.43 ms, total: 1.04 s
Wall time: 1.06 s


In [11]:
%%time
# Построение модели
model = RandomForestRegressor(n_estimators=100, random_state=42, n_jobs=-1)
model.fit(X_train, y_train)

CPU times: user 17min 51s, sys: 4.48 s, total: 17min 55s
Wall time: 11min 14s


RandomForestRegressor(n_jobs=-1, random_state=42)

In [12]:
# Прогнозирование
y_pred = model.predict(X_test)

In [ ]:
# Вычисление RMSE
rmse = mean_squared_error(y_test, y_pred, squared=False)
print(f"RMSE модели: {rmse:.2f}")

RMSE модели: 12.96


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


In [13]:
# Прогнозирование задержки по всем аэропортам
# Обратите внимание, что здесь мы используем тот же набор признаков X
all_predictions = model.predict(X)

In [14]:
# Добавление предсказаний к данным с использованием .loc
flights.loc[:, 'Predicted_ARRIVAL_DELAY'] = all_predictions

In [15]:
# Получение топ-3 аэропортов с минимальными задержками
top_airports = flights.groupby('DESTINATION_AIRPORT')['Predicted_ARRIVAL_DELAY'].mean().nsmallest(3).reset_index()

In [16]:
# Получение информации о названиях аэропортов
top_airport_info = airports[airports['IATA_CODE'].isin(top_airports['DESTINATION_AIRPORT'])]

In [17]:
print("Топ 3 аэропорта прилета с минимальной вероятностью опоздания:")
for index, row in top_airport_info.iterrows():
    print(f"{row['AIRPORT']} ({row['IATA_CODE']})")

Топ 3 аэропорта прилета с минимальной вероятностью опоздания:
Canyonlands Field (CNY)
Dillingham Airport (DLG)
Valdez Airport (VEL)


# Ответы
**Задача 1: Минимальная средняя задержка вылета: Deadhorse Airport (Prudhoe Bay Airport) (SCC) с задержкой 0.15 минут**
****

**Задача 2: Самая пунктуальная авиакомпания на прилет в LAX: Hawaiian Airlines Inc. с задержкой -4.67 минут**
****

**Задача 3: Аэропорт с максимальным временем руления: John F. Kennedy International Airport (New York International Airport) (JFK) с временем 27.01 минут**

****
**Задача 4:**
*   RMSE модели 12.96

**Топ 3 аэропорта прилета с минимальной вероятностью опоздания:**
* Canyonlands Field (CNY)
* Dillingham Airport (DLG)
* Valdez Airport (VEL)